File: practice-2/send-sol.ts
import "dotenv/config";
import {
  Keypair,
  LAMPORTS_PER_SOL,
  PublicKey,
  SystemProgram,
  Transaction,
  clusterApiUrl,
  Connection,
  sendAndConfirmTransaction,
  TransactionInstruction,
} from "@solana/web3.js";

import bs58 from "bs58";

let privateKey = process.env.PRIVATE_KEY;
if (privateKey === undefined) {
  console.log("Add private key to .env!");
  process.exit(1);
}

const asArray = Uint8Array.from(bs58.decode(privateKey));
const sender = Keypair.fromSecretKey(asArray);

const connection = new Connection(clusterApiUrl("devnet"));

console.log("👀 checking balance of", sender.publicKey.toBase58());

const balanceInLamports = await connection.getBalance(sender.publicKey);

const balance = balanceInLamports / LAMPORTS_PER_SOL;

console.log(
  `💲balance of ${sender.publicKey.toBase58()} before sending is ${balance} SOL`,
);
const recipient = new PublicKey("GqdAk31gXguPcLpuzHBTWUzqx6vNfYJFbqT9ewWdwMtw");

const transaction = new Transaction().add(
  SystemProgram.transfer({
    fromPubkey: sender.publicKey,
    toPubkey: recipient,
    lamports: 0.1 * LAMPORTS_PER_SOL,
  }),
);

const memoProgramId = new PublicKey(
  "MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr",
);

transaction.add(
  new TransactionInstruction({
    keys: [{ pubkey: sender.publicKey, isSigner: true, isWritable: false }],
    programId: memoProgramId,
    data: Buffer.from("Some money for you!"),
  }),
);

const signature = await sendAndConfirmTransaction(connection, transaction, [
  sender,
]);
console.log(
  `🎉 Transaction sent: https://explorer.solana.com/tx/${signature}?cluster=devnet`,
);

console.log(
  `💲balance of ${sender.publicKey.toBase58()} after sending is ${
    (await connection.getBalance(sender.publicKey)) / LAMPORTS_PER_SOL
  } SOL`,
);

File: practice-2/rust/src/send_sol.rs
use bs58;
use solana_client::rpc_client::RpcClient;
use solana_sdk::{
    signature::{Keypair, Signer},
    system_instruction,
    transaction::Transaction,
};
use std::env;

pub fn main() {
    let private_key = env::var("PRIVATE_KEY").expect("Add private key to .env!");
    let secret_key = bs58::decode(&private_key)
        .into_vec()
        .expect("Invalid private key");
    let sender = Keypair::from_bytes(&secret_key).expect("Invalid keypair");

    let rpc_url = "https://api.devnet.solana.com";
    let connection = RpcClient::new(rpc_url);

    println!("👀 checking balance of {}", sender.pubkey());

    let balance_in_lamports = connection
        .get_balance(&sender.pubkey())
        .expect("Failed to get balance");
    let balance = balance_in_lamports as f64 / solana_sdk::native_token::LAMPORTS_PER_SOL as f64;

    println!(
        "💲balance of {} before sending is {:.6} SOL",
        sender.pubkey(),
        balance
    );

    let recipient = solana_sdk::pubkey!("GqdAk31gXguPcLpuzHBTWUzqx6vNfYJFbqT9ewWdwMtw");

    let recent_blockhash = connection
        .get_latest_blockhash()
        .expect("Failed to get recent blockhash");

    let lamports = (0.1 * solana_sdk::native_token::LAMPORTS_PER_SOL as f64) as u64;

    let transfer_instruction = system_instruction::transfer(&sender.pubkey(), &recipient, lamports);

    let transaction = Transaction::new_signed_with_payer(
        &[transfer_instruction],
        Some(&sender.pubkey()),
        &[&sender],
        recent_blockhash,
    );

    let signature = connection.send_and_confirm_transaction(&transaction);

    println!(
        "🎉 Transaction sent: https://explorer.solana.com/tx/{}?cluster=devnet",
        signature.unwrap()
    );

    let final_balance_in_lamports = connection
        .get_balance(&sender.pubkey())
        .expect("Failed to get final balance");
    let final_balance =
        final_balance_in_lamports as f64 / solana_sdk::native_token::LAMPORTS_PER_SOL as f64;

    println!(
        "💲balance of {} after sending is {:.6} SOL",
        sender.pubkey(),
        final_balance
    );
}
